File: appjava/src/main/java/com/byoutline/kickmaterial/model/Category.java
package com.byoutline.kickmaterial.model;

import android.graphics.Color;
import android.support.annotation.ColorRes;
import android.support.annotation.DrawableRes;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import com.byoutline.kickmaterial.utils.ColorMixer;
import org.parceler.Parcel;
import org.parceler.ParcelConstructor;

/**
 * @author Pawel Karczewski <pawel.karczewski at byoutline.com> on 2015-03-24
 */
@Parcel(Parcel.Serialization.FIELD)
public class Category {

    public static final int ALL_CATEGORIES_ID = -1;
    public final int categoryId;
    @StringRes
    public final int nameResId;
    @ColorRes
    public final int colorResId;
    @DrawableRes
    public final int drawableResId;
    @ColorRes
    int bgColor = Color.WHITE;

    @ParcelConstructor
    public Category(int categoryId, @StringRes int nameResId, @ColorRes int colorResId, @DrawableRes int drawableResId) {
        this.categoryId = categoryId;
        this.nameResId = nameResId;
        this.colorResId = colorResId;
        this.drawableResId = drawableResId;
    }

    public void setBgColor(@Nullable Integer color) {
        if (color == null) {
            bgColor = Color.WHITE;
        } else {
            bgColor = ColorMixer.mixTwoColors(color, Color.WHITE, 0.20f);
        }
    }

    @ColorRes
    public int getBgColor() {
        return bgColor;
    }
}

File: app/src/main/java/com/byoutline/kickmaterial/model/Category.kt
package com.byoutline.kickmaterial.model

import android.graphics.Color
import android.support.annotation.ColorInt
import android.support.annotation.ColorRes
import android.support.annotation.DrawableRes
import android.support.annotation.StringRes
import com.byoutline.kickmaterial.features.selectcategory.ColorMixer
import paperparcel.PaperParcel
import paperparcel.PaperParcelable

/**
 * @author Pawel Karczewski <pawel.karczewski at byoutline.com> on 2015-03-24
 */
@PaperParcel
class Category(
        val categoryId: Int, @StringRes val nameResId: Int,
        @ColorRes val colorResId: Int,
        @DrawableRes val drawableResId: Int,
        @ColorInt private var bgColor: Int = Color.WHITE
) : PaperParcelable {

    @ColorInt fun getBgColor() = bgColor

    fun setBgColor(color: Int?) {
        bgColor = if (color == null) {
            Color.WHITE
        } else {
            ColorMixer.mixTwoColors(color, Color.WHITE, 0.20f)
        }
    }

    companion object {
        const val ALL_CATEGORIES_ID = -1
        @JvmField val CREATOR = PaperParcelCategory.CREATOR
    }
}



